engine/src/main/java/org/pentaho/di/repository/kdr/delegates/KettleDatabaseRepositoryConnectionDelegate.java

```diff
@@ -0,0 +1,2036 @@
+        return 0; // conversion errors
+      }
+    }
+  }
+
+  public KettleDatabaseRepositoryConnectionDelegate( KettleDatabaseRepository repository, DatabaseMeta databaseMeta ) {
+    super( repository );
+
+    this.databaseMeta = databaseMeta;
+    this.database = new Database( loggingObject, databaseMeta );
+
+    sqlMap = new ConcurrentHashMap<>();
+
+    useBatchProcessing = true; // defaults to true;
+
+    psStepAttributesLookup = null;
+    psStepAttributesInsert = null;
+    psTransAttributesLookup = null;
+    pstmt_entry_attributes = null;
+
+    this.majorVersion = REQUIRED_MAJOR_VERSION;
+    this.minorVersion = REQUIRED_MINOR_VERSION;
+  }
+
+  /**
+   * Connect to the repository
+   */
+  public synchronized void connect() throws KettleException {
+    connect( false, false );
+  }
+
+  public synchronized void connect( boolean no_lookup ) throws KettleException {
+    connect( no_lookup, false );
+  }
+
+  public synchronized void connect( boolean no_lookup, boolean ignoreVersion ) throws KettleException {
+    if ( repository.isConnected() ) {
+      throw new KettleException( "Repository is already connected!" );
+    }
+    try {
+      database.initializeVariablesFrom( null );
+      database.connect();
+      if ( !ignoreVersion ) {
+        verifyVersion();
+      }
+      setAutoCommit( false );
+      repository.setConnected( true );
+      /*
+       * Done lazily now.
+       *
+       * if (!no_lookup) { try { repository.connectionDelegate.setLookupStepAttribute();
+       * repository.connectionDelegate.setLookupTransAttribute();
+       * repository.connectionDelegate.setLookupJobEntryAttribute();
+       * repository.connectionDelegate.setLookupJobAttribute(); } catch (KettleException dbe) { throw new
+       * KettleException("Error setting lookup prep.statements", dbe); } }
+       */
+    } catch ( KettleException e ) {
+      throw new KettleException( "Error connecting to the repository!", e );
+    }
+  }
+
+  /**
+   * Get the required repository version for this version of Kettle.
+   *
+   * @return the required repository version for this version of Kettle.
+   */
+  public static final String getRequiredVersion() {
+    return REQUIRED_MAJOR_VERSION + "." + REQUIRED_MINOR_VERSION;
+  }
+
+  protected void verifyVersion() throws KettleException {
+    RowMetaAndData lastUpgrade = null;
+    String versionTable =
+      databaseMeta.getQuotedSchemaTableCombination( null, KettleDatabaseRepository.TABLE_R_VERSION );
+    try {
+      lastUpgrade = callRead( () ->
+        database.getOneRow( "SELECT "
+          + quote( KettleDatabaseRepository.FIELD_VERSION_MAJOR_VERSION ) + ", "
+          + quote( KettleDatabaseRepository.FIELD_VERSION_MINOR_VERSION ) + ", "
+          + quote( KettleDatabaseRepository.FIELD_VERSION_UPGRADE_DATE ) + " FROM " + versionTable
+          + " ORDER BY " + quote( KettleDatabaseRepository.FIELD_VERSION_UPGRADE_DATE ) + " DESC" ) );
+    } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 214.